File: connector/bacnet/bacnet_connector/connector.py
# Copyright 2026 nexus-gateway contributors
# SPDX-License-Identifier: Apache-2.0

"""BACnet connector: polls / subscribes to BACnet devices and publishes Common Events."""
from __future__ import annotations

import asyncio
import contextlib
import logging
from typing import Any, Callable, Awaitable

import nats
from nats.js import JetStreamContext

from bacnet_connector.config import Config, PointConfig
from bacnet_connector.event import make_event, bacnet_quality

logger = logging.getLogger(__name__)


# BACnetClient is an interface (structural typing) so tests can inject a mock.
class BACnetClient:
    """Minimal interface for BACnet operations needed by the Connector."""

    async def read_property_multiple(
        self,
        address: str,
        device_id: int,
        requests: list[tuple[str, str]],  # [(objid, prop_id), ...]
    ) -> list[tuple[str, float | None, str | None]]:
        """Returns [(objid, value, status), ...]. Raises on transport error."""
        raise NotImplementedError

    async def subscribe_cov(
        self,
        address: str,
        device_id: int,
        obj_id: str,
        callback: Callable[[str, float, str], Awaitable[None]],
        lifetime: int = 300,
    ) -> None:
        """Subscribe to COV notifications for obj_id. Calls callback on each change."""
        raise NotImplementedError

    async def write_property(
        self,
        address: str,
        device_id: int,
        obj_id: str,
        prop_id: str,
        value: float,
        priority: int,
    ) -> None:
        """Write a single property. Raises on device error or transport failure."""
        raise NotImplementedError

    async def close(self) -> None:
        pass


class Connector:
    """Connects to a BACnet device, polls points, and publishes Common Events.

    Dependencies are injected so tests can stub the BACnet and NATS layers.
    """

    def __init__(self, cfg: Config, bacnet: BACnetClient, js: JetStreamContext):
        self._cfg = cfg
        self._bacnet = bacnet
        self._js = js
        self._subject = f"evt.bacnet.{cfg.connector_id}"
        self._point_map = {pt.local_id: pt for pt in cfg.points}
        self._cov_tasks: list[asyncio.Task] = []

    async def run(self, *, stop_event: asyncio.Event | None = None) -> None:
        """Poll all points once, then set up COV subscriptions and keep polling.

        Runs until stop_event is set (or forever if stop_event is None).
        """
        logger.info("bacnet: connector %s starting", self._cfg.connector_id)

        # Initial full poll so we have values immediately on startup.
        await self._poll_all()

        # Subscribe to COV for every point — tasks are cancelled on shutdown.
        # Skipped in poll-only mode: thousands of long-lived COV sessions can
        # overwhelm a device, so large deployments rely on periodic polling alone.
        if self._cfg.cov_enabled:
            for pt in self._cfg.points:
                task = asyncio.create_task(
                    self._subscribe_cov(pt),
                    name=f"cov-{pt.local_id}",
                )
                self._cov_tasks.append(task)

        # Periodic poll loop — keeps values fresh even when COV subscriptions lapse.
        # Uses stop_event.wait() with a timeout so shutdown is immediate.
        try:
            while True:
                if stop_event is not None:
                    try:
                        await asyncio.wait_for(
                            stop_event.wait(),
                            timeout=self._cfg.poll_interval,
                        )
                        break  # stop_event fired
                    except asyncio.TimeoutError:
                        pass  # poll interval elapsed, continue
                else:
                    await asyncio.sleep(self._cfg.poll_interval)
                await self._poll_all()
        finally:
            for t in self._cov_tasks:
                t.cancel()
            await asyncio.gather(*self._cov_tasks, return_exceptions=True)
            self._cov_tasks.clear()
            await self._bacnet.close()
            logger.info("bacnet: connector %s stopped", self._cfg.connector_id)

    async def _poll_all(self) -> None:
        """Poll all points and publish events.

        Points are read in chunks of ``rpm_chunk_size`` rather than a single
        ReadPropertyMultiple: a response for many points can overflow the device's
        APDU, and devices without segmentation reject the whole request with
        "segmentation-not-supported". A failed chunk is logged and skipped so the
        remaining chunks still publish.
        """
        if not self._cfg.points:
            return

        requests = [(pt.local_id, "presentValue") for pt in self._cfg.points]
        chunk_size = self._cfg.rpm_chunk_size
        for start in range(0, len(requests), chunk_size):
            chunk = requests[start:start + chunk_size]
            try:
                results = await self._bacnet.read_property_multiple(
                    self._cfg.bacnet_address,
                    self._cfg.bacnet_device_id,
                    chunk,
                )
            except Exception as exc:
                logger.warning(
                    "bacnet: poll chunk [%d:%d] failed: %s",
                    start, start + len(chunk), exc,
                )
                continue

            for obj_id, value, status in results:
                pt = self._point_map.get(obj_id)
                if pt is None or value is None:
                    continue
                await self._publish(pt, value, bacnet_quality(status))

    async def _subscribe_cov(self, pt: PointConfig) -> None:
        """Subscribe to COV for a single point and publish events on each change."""
        async def on_cov(obj_id: str, value: float, status: str) -> None:
            await self._publish(pt, value, bacnet_quality(status))

        while True:
            try:
                await self._bacnet.subscribe_cov(
                    self._cfg.bacnet_address,
                    self._cfg.bacnet_device_id,
                    pt.local_id,
                    callback=on_cov,
                )
                return  # subscription active — callbacks will fire
            except asyncio.CancelledError:
                return
            except Exception as exc:
                logger.warning(
                    "bacnet: COV subscribe failed for %s, retrying in 30s: %s",
                    pt.local_id, exc,
                )
                await asyncio.sleep(30)

    async def _publish(self, pt: PointConfig, value: float, quality: str) -> None:
        data = make_event(
            connector_id=self._cfg.connector_id,
            local_id=pt.local_id,
            device_ref=pt.device_ref,
            value=value,
            unit=pt.unit,
            quality=quality,
        )
        try:
            await self._js.publish(self._subject, data)
            logger.debug("bacnet: published %s=%s quality=%s", pt.local_id, value, quality)
        except Exception as exc:
            logger.warning("bacnet: NATS publish failed — withholding for %s: %s", pt.local_id, exc)

File: connector/opcua/src/main/java/io/nexus/gateway/opcua/Connector.java
// Copyright 2026 nexus-gateway contributors
// SPDX-License-Identifier: Apache-2.0

package io.nexus.gateway.opcua;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/** Connects to an OPC-UA server and publishes Common Events via the injected publisher. */
public class Connector {

    @FunctionalInterface
    public interface Publisher {
        void publish(String subject, byte[] data) throws Exception;
    }

    private static final Logger log = LoggerFactory.getLogger(Connector.class);
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Config cfg;
    private final OpcUaClientFacade client;
    private final Publisher publisher;
    private final Map<String, PointConfig> pointMap;
    private final List<String> nodeIds;
    private final String subject;
    private final CountDownLatch stopLatch = new CountDownLatch(1);

    public Connector(Config cfg, OpcUaClientFacade client, Publisher publisher) {
        this.cfg = cfg;
        this.client = client;
        this.publisher = publisher;
        this.pointMap = cfg.points().stream()
            .collect(Collectors.toMap(PointConfig::localId, p -> p));
        this.nodeIds = cfg.points().stream().map(PointConfig::localId).toList();
        this.subject = "evt.opcua." + cfg.connectorId();
    }

    /**
     * Connect, poll once, subscribe, then re-poll every {@code pollInterval}
     * seconds until stop() is called. The periodic re-poll is a freshness floor
     * alongside the change-driven subscription (#110): with static server values
     * the subscription fires nothing, so polling guarantees every point is
     * refreshed at least once per interval.
     */
    public void run() throws Exception {
        log.info("opcua: connector {} starting, endpoint={}", cfg.connectorId(), cfg.opcuaEndpoint());
        client.connect();

        // Log browse results for Point List authoring (AC: browse logged locally)
        try {
            Map<String, String> nodes = client.browse("i=85"); // Objects folder
            log.info("opcua: browse found {} nodes under Objects", nodes.size());
            nodes.forEach((id, name) -> log.info("  {} => {}", id, name));
        } catch (Exception ex) {
            log.warn("opcua: browse failed (non-fatal): {}", ex.getMessage());
        }

        try {
            // Initial poll
            pollAll();

            // Subscribe to monitored items for all configured points
            if (!nodeIds.isEmpty()) {
                client.subscribe(nodeIds, this::onValue);
            }

            log.info("opcua: connector {} subscribed to {} points, re-poll every {}s",
                cfg.connectorId(), nodeIds.size(), cfg.pollInterval());

            // Periodic re-poll backstop alongside the subscription (#110).
            // await() returns true once stop() fires; false on timeout (interval
            // elapsed) → re-poll. At least 1 ms to avoid a busy loop.
            long intervalMs = Math.max(1L, (long) (cfg.pollInterval() * 1000));
            while (!stopLatch.await(intervalMs, TimeUnit.MILLISECONDS)) {
                pollAll();
            }
        } finally {
            client.close();
            log.info("opcua: connector {} stopped", cfg.connectorId());
        }
    }

    public void stop() {
        stopLatch.countDown();
    }

    private void onValue(String nodeId, OpcValue opcValue) {
        PointConfig pt = pointMap.get(nodeId);
        if (pt == null) return;
        Double value = opcValue.toDouble();
        if (value == null) {
            log.debug("opcua: skipping non-numeric value for {}", nodeId);
            return;
        }
        publish(pt, value, opcValue.quality().toCommonQuality());
    }

    private void pollAll() {
        if (nodeIds.isEmpty()) return;
        try {
            Map<String, OpcValue> results = client.read(nodeIds);
            results.forEach((nodeId, opcValue) -> {
                PointConfig pt = pointMap.get(nodeId);
                if (pt == null) return;
                Double value = opcValue.toDouble();
                if (value == null) return;
                publish(pt, value, opcValue.quality().toCommonQuality());
            });
        } catch (Exception ex) {
            log.warn("opcua: poll failed: {}", ex.getMessage());
        }
    }

    private void publish(PointConfig pt, double value, String quality) {
        try {
            byte[] data = MAPPER.writeValueAsBytes(
                CommonEvent.now(cfg.connectorId(), pt.localId(), pt.deviceRef(), value, pt.unit(), quality)
            );
            publisher.publish(subject, data);
            log.debug("opcua: published {}={} quality={}", pt.localId(), value, quality);
        } catch (Exception ex) {
            log.warn("opcua: publish failed for {}: {}", pt.localId(), ex.getMessage());
        }
    }
}
